query_optimizer/expressions/SubqueryExpression.hpp

```diff
@@ -0,0 +1,126 @@
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_QUERY_OPTIMIZER_EXPRESSIONS_SUBQUERY_EXPRESSION_HPP_
+#define QUICKSTEP_QUERY_OPTIMIZER_EXPRESSIONS_SUBQUERY_EXPRESSION_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "query_optimizer/expressions/AttributeReference.hpp"
+#include "query_optimizer/expressions/Expression.hpp"
+#include "query_optimizer/expressions/ExpressionType.hpp"
+#include "query_optimizer/expressions/Scalar.hpp"
+#include "query_optimizer/logical/Logical.hpp"
+#include "query_optimizer/OptimizerTree.hpp"
+#include "utility/Macros.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+
+class Type;
+
+namespace optimizer {
+namespace expressions {
+
+/** \addtogroup OptimizerExpressions
+ *  @{
+ */
+
+class SubqueryExpression;
+typedef std::shared_ptr<const SubqueryExpression> SubqueryExpressionPtr;
+
+/**
+ * @brief A subquery used in an expression.
+ */
+class SubqueryExpression : public Scalar {
+ public:
+  ExpressionType getExpressionType() const override {
+    return ExpressionType::kSubqueryExpression;
+  }
+
+  std::string getName() const override { return "SubqueryExpression"; }
+
+  const Type& getValueType() const override {
+    return output_attribute_->getValueType();
+  }
+
+  bool isConstant() const override {
+    return output_attribute_->isConstant();
+  }
+
+  /**
+   * @return The referenced logical subquery node.
+   */
+  const logical::LogicalPtr& subquery() const {
+    return subquery_;
+  }
+
+  std::vector<AttributeReferencePtr> getReferencedAttributes() const override;
+
+  ExpressionPtr copyWithNewChildren(
+      const std::vector<ExpressionPtr> &new_children) const override {
+    DCHECK(new_children.empty());
+    return Create(subquery_);
+  }
+
+  ::quickstep::Scalar* concretize(
+      const std::unordered_map<ExprId, const CatalogAttribute*>& substitution_map) const override;
+
+  /**
+   * @brief Creates a subquery expression.
+   * @note This expression can only be used in a logical plan.
+   *
+   * @param subquery The logical subquery node.
+   * @return An immutable SubqueryExpression.
+   */
+  static SubqueryExpressionPtr Create(const logical::LogicalPtr& subquery) {
+    return SubqueryExpressionPtr(new SubqueryExpression(subquery));
+  }
+
+ protected:
+  void getFieldStringItems(
+      std::vector<std::string> *inline_field_names,
+      std::vector<std::string> *inline_field_values,
+      std::vector<std::string> *non_container_child_field_names,
+      std::vector<OptimizerTreeBaseNodePtr> *non_container_child_fields,
+      std::vector<std::string> *container_child_field_names,
+      std::vector<std::vector<OptimizerTreeBaseNodePtr>> *container_child_fields) const override;
+
+ private:
+  explicit SubqueryExpression(const logical::LogicalPtr& subquery)
```

Comment:
What do you think if we apply the reference sigil standard here?

`const logical::LogicalPtr &subquery`.